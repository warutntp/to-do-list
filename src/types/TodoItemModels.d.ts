export module TodoItemModels {
  export interface TodoItemModel {
    slug: string;
    title: string;
    description?: string;
    completed: boolean;
  }
}
