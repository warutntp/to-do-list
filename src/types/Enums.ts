export enum FilterTodoType {
  Completed = "completed",
  Incomplete = "incomplete",
  All = "all",
}
