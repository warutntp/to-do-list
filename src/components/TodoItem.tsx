import React, { useState } from "react";
import { TodoItemModels } from "../types/TodoItemModels";
import { useDispatch } from "react-redux";
import { toggleTodo, removeTodo } from "../store/slices/todosSlice";
import { Modal, Button } from "react-bootstrap";

interface TodoItemProps {
  todo: TodoItemModels.TodoItemModel;
  index: number;
}

const TodoItem: React.FC<TodoItemProps> = ({ todo, index }) => {
  const dispatch = useDispatch();
  const [showDeletePopup, setShowDeletePopup] = useState(false);

  const handleToggleTodo = (slug: string) => {
    dispatch(toggleTodo(slug));
  };

  const handleRemoveTodo = (slug: string) => {
    dispatch(removeTodo(slug));
    setShowDeletePopup(false);
  };

  const handleShow = () => setShowDeletePopup(true);
  const handleClose = () => setShowDeletePopup(false);

  return (
    <>
      <tr>
        <td className="col-1 text-center">{index + 1}</td>
        <td className="col-3 text-wrap-break">{todo.title}</td>
        <td className="col-4 text-wrap-break">{todo.description}</td>
        <td className="col-2">
          <div className="form-check form-switch d-flex justify-content-center">
            <input
              className="form-check-input"
              type="checkbox"
              id={todo.slug}
              checked={todo.completed}
              onChange={() => handleToggleTodo(todo.slug)}
            />
          </div>
        </td>
        <td className="text-center">
          <button
            type="button"
            className="btn py-0"
            onClick={handleShow}
            title="Delete"
          >
            <i className="bi bi-trash bi-danger" />
          </button>
        </td>
      </tr>
      <Modal show={showDeletePopup} onHide={handleClose} centered>
        <Modal.Header closeButton>
          <Modal.Title>Confirm Delete</Modal.Title>
        </Modal.Header>
        <Modal.Body>{`Are you sure you want to delete "${todo.title}" ?`}</Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={handleClose}>
            Cancel
          </Button>
          <Button variant="danger" onClick={() => handleRemoveTodo(todo.slug)}>
            Delete
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
};

export default TodoItem;
