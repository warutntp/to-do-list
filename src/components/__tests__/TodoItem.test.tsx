import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import { TodoItemModels } from "../../types/TodoItemModels";
import TodoItem from "../TodoItem";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

const mockStore = configureStore([]);
const store = mockStore({});

const todo: TodoItemModels.TodoItemModel = {
  slug: "1",
  title: "Test Todo",
  description: "Test description",
  completed: false,
};

describe("TodoItem", () => {
  it("should render TodoItem correctly", () => {
    render(
      <Provider store={store}>
        <table>
          <tbody>
            <TodoItem todo={todo} index={0} />
          </tbody>
        </table>
      </Provider>
    );

    expect(screen.getByText("Test Todo")).toBeInTheDocument();
    expect(screen.getByText("Test description")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /delete/i })).toBeInTheDocument();
  });

  it("should open confirmation modal when delete button is clicked", () => {
    render(
      <Provider store={store}>
        <table>
          <tbody>
            <TodoItem todo={todo} index={0} />
          </tbody>
        </table>
      </Provider>
    );

    fireEvent.click(screen.getByRole("button", { name: /delete/i }));
    expect(
      screen.getByText(`Are you sure you want to delete "${todo.title}" item?`)
    ).toBeInTheDocument();
  });
});
