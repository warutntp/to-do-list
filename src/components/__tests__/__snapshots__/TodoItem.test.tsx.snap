// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem should match snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="col-1 text-center"
        >
          1
        </td>
        <td
          class="col-3 text-wrap-break"
        >
          Test Todo
        </td>
        <td
          class="col-4 text-wrap-break"
        >
          Test description
        </td>
        <td
          class="col-2"
        >
          <div
            class="form-check form-switch d-flex justify-content-center"
          >
            <input
              class="form-check-input"
              id="1"
              type="checkbox"
            />
          </div>
        </td>
        <td
          class="text-center"
        >
          <button
            class="btn py-0"
            title="Delete"
            type="button"
          >
            <i
              class="bi bi-trash bi-danger"
            />
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
